File: services/auth-service/src/main/java/com/medmatch/auth/entity/AuditLog.java
package com.medmatch.auth.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.LocalDateTime;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

// NOTE: audit_logs is a shared table, physically owned (migrated) by
// auth-service via Flyway, but also written directly by the AI service
// (see services/ai-service/app/models/audit_log.py and
// app/repositories/audit_log_repository.py). The physical table has
// 12 columns; this entity intentionally maps only the 8 that
// auth-service itself uses.
//
// The 4 columns below are NOT mapped here and are always NULL on rows
// written by auth-service. They are nullable, ai-service-only columns,
// populated exclusively by the AI service from JWT claims at write time:
//   - performed_by_username (varchar(255))
//   - performed_by_role     (varchar(50))
//   - hospital_id           (uuid)
//   - hospital_name         (varchar(255))
//
// This is a deliberate minimal-compatibility choice, not an oversight.
// Hibernate's ddl-auto=validate only checks columns this entity maps,
// so leaving these unmapped is safe as long as `action` and
// `resource_type` declare the same lengths as the physical column
// (see V4__create_audit_logs_table.sql). Full Java-side parity
// (mapping these fields and extending AuditService.createAuditLog)
// is a possible future change, not implemented in this phase.
@Entity
@Table(name = "audit_logs")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class AuditLog {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @JdbcTypeCode(SqlTypes.UUID)
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    // Stored as a raw UUID column, not a @ManyToOne to User — keeps this
    // entity independent per the design requirement, and means audit
    // records survive even if the referenced User row is later deleted.
    @JdbcTypeCode(SqlTypes.UUID)
    @Column(name = "performed_by_id")
    private UUID performedById;

    @Column(name = "action", nullable = false, length = 100)
    private String action;

    @Column(name = "resource_type", length = 100)
    private String resourceType;

    @Column(name = "resource_id")
    private UUID resourceId;

    @Column(name = "ip_address")
    private String ipAddress;

    @Column(name = "details", columnDefinition = "TEXT")
    private String details;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
}

File: services/ai-service/app/models/audit_log.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class AuditLog(Base):
    __tablename__ = "audit_logs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    action: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    performed_by_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
        index=True,
    )

    performed_by_username: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    performed_by_role: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    hospital_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
        index=True,
    )

    hospital_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    resource_type: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    resource_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
    )

    details: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )